### ISSUE TITLE
Improve understandability of AGW in NMS - health score

### ISSUE THREAD

**Issue by wallyrb** [magma/magma#8465]
### Your Environment (If applicable)

- **Affected Component:** AGW, Orchestrator, NMS

### Describe the problem

**Description**
Every time we access NMS, AGW Health(under Equipment) only shows connection status between Orc8r and AGW. This doesn't provide enough information about current issues happening in AGW. This can lead to miss important issues in AGW. For example, high memory in a specific service, service not active or have frequent restarts, some interfaces not up, congestion, etc. 

**Additional context**
Below shows picture of AGW health, which good means AGW has checkin with Orc8r, and bad means connection between AGW and Orc8r is not OK. 

![image](https://user-images.githubusercontent.com/37117037/128155335-18f9aaf0-70d3-4079-8124-00e6b516f53a.png)

**Desired timelines**
2021 H2

### Describe your solution
**Description**
Split the current AGW Health in two parameters:
- Health score: Collect different signals from AGW and aggregate those in a single AGW health score that can range from 0-100. 
- Availability: We can add if the AGW is offline or online depending if Orc8r connection is healthy(current Health approach)

We can show this information with a single score for aggregated view and heatmap for every AGW and all signals for more granularity.

**Additional context**
We can aggregate several signals from AGW and aggregate in one single value. Each signal will have a different weight, that should impact the overall score. Weight should be customizable for magma user needs. For example, mme service not active, should imply a poor score in AGW health score. Example of signals: 

- service status(active, inactive)
- service memory usage
- service disk usage
- service cpu usage
- Number of service restarts in the last 24 hours
- Congestion (ITTI bus, delay in response)
- Number of registered users: mobility_cli.py get_subscriber_table
- Interfaces health(ip addr)
- pipelined_cli.py debug display_flows
- Checkin success volume(last 5 minutes, last hour)
- connectivity/delay/packet loss. (ping google.com -I eth0 -c 5)
- eNB connection health(sctp heartbeat messages)

**Comment by krramasw** [comment]
First of all, it is a great idea to introduce a health score metric. However, would like to point out no two customers would see the summarized health score in the same way and many would want to tinker or customize the weightage to suit their need.

**Comment by wallyrb** [comment]
> First of all, it is a great idea to introduce a health score metric. However, would like to point out no two customers would see the summarized health score in the same way and many would want to tinker or customize the weightage to suit their need.

You are right. I also think weight should be customizable. Updated the issue considering this. Thanks @krramasw !

**Comment by jingtien** [comment]
Wally, Jared, Iris, and Jing discussed this project today.  Recap of next steps:

_Phase 1: Research_
Wally/Jared will work with XFN teams to research the following
**Health score**
- Determine which signals (metrics, etc) should be used in a health score
- Figure out the granularity needed in a health score (1-100?, green/orange/red buckets?)
- How should each signal be weighted?  ex. is service status more important than service disk usage?

**For each signal (metric)**
- Figure out the granularity of each signal.  Score of 1-100?  Green/orange/red buckets?  Ex. CPU usage red/poor
- The thresholds for each metric (what is considered good or poor?)

**Customization**
- Do partners want to change the weight of different signals in a health score?
- Do partners need to change the thresholds for an individual signal?  Ex. we say that CPU usage of 50% is poor, but they want to say that 50% is ok

_Phase 2: Exploration_
- Wally, Jared, Iris, Jing to review research together
- Iris/Jing to explore how to visualize and explain a health score in the UI
- Iris/Jing to determine entry points for health score in the UI (dashboard, tables, etc)

**Comment by jingtien** [comment]
I can't find a way to edit the description so I'll add the impact as a comment so Lindsey can help us with prioritization.

**Impact**
- Summarizes lots of data to make it easier to consume, which then makes it faster for users to recognize issues and faster to fix them
- Gives users with lower technical expertise the ability to understand the status of the AGW while still allowing them to drill into details and scale their technical expertise

**Reach**
High, allows users to understand the overall AGW health without looking at 10 different metrics in several places

**Confidence**
Medium/High, the Terragraph NMS is also exploring this same problem and solution.  Users don't have an enough expertise to understand low level details, they want things rolled up into a higher level understanding.

**Eng Effort**
tbd

**Comment by jingtien** [comment]
We're merging with this task with this one [#8762](https://app.zenhub.com/workspaces/magma-5fac75d3e2cd890011f1677a/issues/magma/magma/8762)

Notes from #8762
The Magma AGW has a number of CLI tools that are critical for monitoring, setup, and debug. To access these tools and use them reliably, requires both SSH access to the AGW, often requires explicit knowledge of these tools, and even some usage info that is available mainly to the developers of these tools.

This represents a large usability gap, as users are not able to use the tools that we have built.

Task
- Allow NMS access to all AGW CLI tools
- Provide an easy user interface for all these CLI tools
- Provide good feature discoverability of these CLI tools

**Comment by jaredmullane** [comment]
Another option would be instead of a column with a 0-100 score would be to have a column called "Issues" and if health is bad it would show the number of failing signals and a description. 

For example "Issues: 2 [service disk usage, congestion]" and then a user could check the FAQ documentation for how to troubleshoot those issues. cc: @wallyrb

**Comment by jingtien** [comment]
[1-pager doc](https://docs.google.com/document/d/1fz2vKBxSOv_punrKdFbUm7yvv6gRygQiUyjDoSHaqHk/edit?usp=sharing) summarizing problem and approach

**Comment by jingtien** [comment]
Email jingtien@fb.com for access to docs.

References
- [Description of problem and approach](https://docs.google.com/document/d/1fz2vKBxSOv_punrKdFbUm7yvv6gRygQiUyjDoSHaqHk/edit?usp=sharing)
- View [design mocks](https://www.figma.com/file/46OyCpHAE7AOgp9RKrgYfv/AGW-Health?node-id=608%3A36190)
- How [health status/score is calculated](https://docs.google.com/document/d/1ADD9QH1MD1cRG94pOCfIyTwH4IFFcUcZof_p7RqwYRU/edit?usp=sharing)

**Comment by p00jaY** [comment]
Hi @jingtien ,

Any update when the fix for improving the AGW Bad state will be available. Can you please share your plan for the fix.

Thanks and Regards,
Pooja Yadav

**Comment by jingtien** [comment]
Hi @p00jaY, it looks like Thomas Schmitt has assigned himself this feature so I will defer to him for the timeline.  @thmsschmitt are there any general estimates or prioritizations you can share?

**Comment by thmsschmitt** [comment]
Hi, @p00jaY there is no timeline for this yet. But I will look into #12004 soon if that helps.

**Comment by shahrahman-fb** [comment]
Closing this issue as part of a repository cleanup effort. This issue has had no activity for over 2 years.

If this issue is still relevant, please reopen it or file a new issue with updated information. We are actively maintaining the project again and want to focus on current, actionable issues.

See: #15833 (Phase 1 cleanup), #15828 (tracking issue)
